### ISSUE TITLE
Adapt voxel grid with signed distance function to surface extraction of poisson reconstruction

### ISSUE THREAD

**Issue by Dennis199607** [mkazhdan/PoissonRecon#157]
Hi community,

Following problem: I have a voxel grid (not as an octree just a regular grid) in which every point of a voxel has the signed distance to the surface. I need to convert my voxel grid to the octree, which is used in the Screened Poisson reconstruction in order to use the adapted marching cubes algorithm used in Screened Poisson Reconstruction. I tried to debug the poisson reconstruction step by step to see how the data for marching cubes in Screened Poisson Reconstruction is set up, but it is very hard to read and understand.

What would be the easiest way to get this done? My current idea is to take the points (the positions with a pseudo normal, because I have no normals) of my voxel grid as input. The GetMCIsoSurface method takes parameters densityWeights, solution and isoValue. For signed distance function isoValue would be zero, but I do not know what to do with densityWeights and solution, since this will not be correctly set up if I do not provide correct normals.

Can someone help me who is more proficient with the code and the theory all behind that? Big thanks in advance!

**Comment by mkazhdan** [comment]
Given that you have the signal represented over a regular grid, why not just use code for Marching Cubes on a regular grid. Such code can be found, for example, here:
     https://github.com/mkazhdan/IsoSurfaceExtraction

**Comment by Dennis199607** [comment]
The normal marching cubes iterating over every voxel and creating triangles is already implemented in my code. But one of my  tasks is to extend marching cubes algorithm adapted to octrees like in your code for Screened Poisson Reconstruction. My supervisor told me to build with your code an octree from my voxel grid, that seems to be pretty difficult for me and is my problem :/

**Comment by mkazhdan** [comment]
Ah. I see. That would be hard. Both in terms of getting an octree representation out of your voxel grid and in terms of extracting the iso-surface from it. I am afraid I don't see an easy way to do it. It's possible that the code at:
http://www.cs.jhu.edu/~misha/Code/IsoOctree/
would be helpful.

**Comment by Dennis199607** [comment]
I took a look at the code of IsoOctree and it really helped me to progress. They actually take an input mesh and calculating for every corner of their Octree the distance to the mesh and run marching cubes over the Octree. I just need to substitute my distance values of my grid with their distance values of their Octree, should be doable since I just need to adapt their setter function. The only thing I am not sure about is if they use same marching cubes algorithm adapted to octrees like in your code of screened poisson reconstruction? Nevertheless really big thx mkahzdan.

**Comment by mkazhdan** [comment]
Yep. They both use the method described in the 2007 paper "Unconstrained Isosurface Extraction on Arbitrary Octrees"
